### ISSUE TITLE
Conflict with Rank Math: wp.data.subscribe callbacks block editor initialization

### ISSUE THREAD

**Issue by stevejburge** [publishpress/publishpress-checklists#1169]
## Summary

When Rank Math SEO is active alongside PublishPress Checklists, Rank Math's SEO panel in the block editor loads stuck at 0/100 and fails to render its options. Switching browser tabs and returning forces a re-render that fixes it. Disabling Checklists resolves the issue entirely.

**Reported by user:** https://wordpress.org/support/topic/conflict-with-rank-math-9/

---

## Root Cause

There are two compounding issues in `modules/checklists/assets/js/meta-box.js`.

### 1. `wp.data.subscribe` callbacks are registered immediately, bypassing the Rank Math delay

When Rank Math is detected, `PP_Checklists.init()` is correctly delayed by 3 seconds:

```js
} else if (typeof rankMath !== 'undefined') {
    setTimeout(function () {
      PP_Checklists.init();
    }, 3000);
```

However, approximately 8 `wp.data.subscribe` calls are registered **outside** of `init()`, at the IIFE top level. They fire on every store update from the moment the script loads, regardless of the delay:

- Line ~999 — `featured_image_alt`
- Line ~1071 — `featured_image_caption`
- Line ~1676 — `words_count`
- Line ~1807 — `internal_links`
- Line ~1941 — `external_links`
- Line ~2074 — `image_alt`
- Line ~2230 — `validate_links`
- Line ~2359 — `image_alt_count`

### 2. Those subscribe callbacks perform heavy synchronous work on every store update

During block editor initialization the WordPress data store fires dozens of rapid state changes. Each one causes all PPCH subscribers to run synchronously:

- Access `wp.data.select('core/editor').getEditedPostAttribute('content')`
- Run regex operations over the full post content (link extraction, alt-text parsing)
- Run `document.querySelector` calls for block list-view DOM elements

Because `wp.data.subscribe` is synchronous, this blocks the main JS thread during the exact window in which Rank Math's React panel is mounting and fetching its initial SEO score. The panel renders at 0/100 and cannot proceed.

**Why the tab-switch workaround works:** After the initial load burst, the data store settles. When focus returns to the tab, Rank Math re-renders without contention and completes normally.

---

## Fix

**Option A (recommended):** Move all `wp.data.subscribe` registrations inside `init()` so they also respect the 3-second delay when Rank Math is present. This is the minimal, targeted fix.

**Option B:** Add debouncing to all subscribe callbacks. The `internal_links` and `external_links` handlers already wrap their work in `setTimeout(100/150)`, but `image_alt`, `image_alt_count`, `validate_links`, and `words_count` run fully synchronously on every store tick and should be debounced similarly.

Ideally both options would be applied together.

**Comment by alexggr** [comment]
Hi, commenting here as requested on this issue's Wordpress support thread!

**Comment by rizaardiyanto1412** [comment]
@alexggr Please try this version and let me know if the issue persists: https://drive.google.com/file/d/1cVHlf86aTH6PF5LbtJywektMqJtFQ8D2/view?usp=sharing

**Comment by alexggr** [comment]
@rizaardiyanto1412 The issue is still happening, I'm afraid.

**Comment by rizaardiyanto1412** [comment]
@alexggr I’m having trouble reproducing this on my end, so it’s hard to confirm whether the patch I’m working on actually fixes it.

Would you be willing to share a full backup of the affected site so we can restore it locally, work on the patch, and verify the fix directly here instead of sending patches back and forth?

If you are down, you could use plugin All In One WP Migration, then export it, send that export file here through Google Drive or something and we will restore it from here
